File: cli/src/tests/language_test.rs
use tree_sitter::{self, Parser};

use super::helpers::fixtures::get_language;

#[test]
fn test_lookahead_iterator() {
    let mut parser = Parser::new();
    let language = get_language("rust");
    parser.set_language(&language).unwrap();

    let tree = parser.parse("struct Stuff {}", None).unwrap();

    let mut cursor = tree.walk();

    assert!(cursor.goto_first_child()); // struct
    assert!(cursor.goto_first_child()); // struct keyword

    let next_state = cursor.node().next_parse_state();
    assert_ne!(next_state, 0);
    assert_eq!(
        next_state,
        language.next_state(cursor.node().parse_state(), cursor.node().grammar_id())
    );
    assert!((next_state as usize) < language.parse_state_count());
    assert!(cursor.goto_next_sibling()); // type_identifier
    assert_eq!(next_state, cursor.node().parse_state());
    assert_eq!(cursor.node().grammar_name(), "identifier");
    assert_ne!(cursor.node().grammar_id(), cursor.node().kind_id());

    let expected_symbols = ["//", "/*", "identifier", "line_comment", "block_comment"];
    let mut lookahead = language.lookahead_iterator(next_state).unwrap();
    assert_eq!(*lookahead.language(), language);
    assert!(lookahead.iter_names().eq(expected_symbols));

    lookahead.reset_state(next_state);
    assert!(lookahead.iter_names().eq(expected_symbols));

    lookahead.reset(&language, next_state);
    assert!(lookahead
        .map(|s| language.node_kind_for_id(s).unwrap())
        .eq(expected_symbols));
}

#[test]
fn test_lookahead_iterator_modifiable_only_by_mut() {
    let mut parser = Parser::new();
    let language = get_language("rust");
    parser.set_language(&language).unwrap();

    let tree = parser.parse("struct Stuff {}", None).unwrap();

    let mut cursor = tree.walk();

    assert!(cursor.goto_first_child()); // struct
    assert!(cursor.goto_first_child()); // struct keyword

    let next_state = cursor.node().next_parse_state();
    assert_ne!(next_state, 0);

    let mut lookahead = language.lookahead_iterator(next_state).unwrap();
    let _ = lookahead.next();

    let mut names = lookahead.iter_names();
    let _ = names.next();
}

#[test]
fn test_symbol_metadata_checks() {
    let language = get_language("rust");
    for i in 0..language.node_kind_count() {
        let sym = i as u16;
        let name = language.node_kind_for_id(sym).unwrap();
        match name {
            "_type"
            | "_expression"
            | "_pattern"
            | "_literal"
            | "_literal_pattern"
            | "_declaration_statement" => assert!(language.node_kind_is_supertype(sym)),

            "_raw_string_literal_start"
            | "_raw_string_literal_end"
            | "_line_doc_comment"
            | "_error_sentinel" => assert!(!language.node_kind_is_supertype(sym)),

            "enum_item" | "struct_item" | "type_item" => {
                assert!(language.node_kind_is_named(sym));
            }

            "=>" | "[" | "]" | "(" | ")" | "{" | "}" => {
                assert!(language.node_kind_is_visible(sym));
            }

            _ => {}
        }
    }
}

#[test]
fn test_supertypes() {
    let language = get_language("rust");
    let supertypes = language.supertypes();

    if language.abi_version() < 15 {
        return;
    }

    assert_eq!(supertypes.len(), 5);
    assert_eq!(
        supertypes
            .iter()
            .filter_map(|&s| language.node_kind_for_id(s))
            .map(|s| s.to_string())
            .collect::<Vec<String>>(),
        vec![
            "_expression",
            "_literal",
            "_literal_pattern",
            "_pattern",
            "_type"
        ]
    );

    for &supertype in supertypes {
        let mut subtypes = language
            .subtypes_for_supertype(supertype)
            .iter()
            .filter_map(|symbol| language.node_kind_for_id(*symbol))
            .collect::<Vec<&str>>();
        subtypes.sort_unstable();
        subtypes.dedup();

        match language.node_kind_for_id(supertype) {
            Some("_literal") => {
                assert_eq!(
                    subtypes,
                    &[
                        "boolean_literal",
                        "char_literal",
                        "float_literal",
                        "integer_literal",
                        "raw_string_literal",
                        "string_literal"
                    ]
                );
            }
            Some("_pattern") => {
                assert_eq!(
                    subtypes,
                    &[
                        "_",
                        "_literal_pattern",
                        "captured_pattern",
                        "const_block",
                        "generic_pattern",
                        "identifier",
                        "macro_invocation",
                        "mut_pattern",
                        "or_pattern",
                        "range_pattern",
                        "ref_pattern",
                        "reference_pattern",
                        "remaining_field_pattern",
                        "scoped_identifier",
                        "slice_pattern",
                        "struct_pattern",
                        "tuple_pattern",
                        "tuple_struct_pattern",
                    ]
                );
            }
            Some("_type") => {
                assert_eq!(
                    subtypes,
                    &[
                        "abstract_type",
                        "array_type",
                        "bounded_type",
                        "dynamic_type",
                        "function_type",
                        "generic_type",
                        "macro_invocation",
                        "metavariable",
                        "never_type",
                        "pointer_type",
                        "primitive_type",
                        "reference_type",
                        "removed_trait_bound",
                        "scoped_type_identifier",
                        "tuple_type",
                        "type_identifier",
                        "unit_type"
                    ]
                );
            }
            _ => {}
        }
    }
}

File: lib/binding_web/test/language.test.ts
import { describe, it, expect, beforeAll, afterAll } from 'vitest';
import helper from './helper';
import type { LookaheadIterator, Language } from '../src';
import { Parser } from '../src';

let JavaScript: Language;
let Rust: Language;

describe('Language', () => {
  beforeAll(async () => ({ JavaScript, Rust } = await helper));

  describe('.name, .version', () => {
    it('returns the name and version of the language', () => {
      expect(JavaScript.name).toBe('javascript');
      expect(JavaScript.abiVersion).toBe(15);
    });
  });

  describe('.fieldIdForName, .fieldNameForId', () => {
    it('converts between the string and integer representations of fields', () => {
      const nameId = JavaScript.fieldIdForName('name');
      const bodyId = JavaScript.fieldIdForName('body');

      expect(nameId).toBeLessThan(JavaScript.fieldCount);
      expect(bodyId).toBeLessThan(JavaScript.fieldCount);
      expect(JavaScript.fieldNameForId(nameId!)).toBe('name');
      expect(JavaScript.fieldNameForId(bodyId!)).toBe('body');
    });

    it('handles invalid inputs', () => {
      expect(JavaScript.fieldIdForName('namezzz')).toBeNull();
      expect(JavaScript.fieldNameForId(-3)).toBeNull();
      expect(JavaScript.fieldNameForId(10000)).toBeNull();
    });
  });

  describe('.idForNodeType, .nodeTypeForId, .nodeTypeIsNamed', () => {
    it('converts between the string and integer representations of a node type', () => {
      const exportStatementId = JavaScript.idForNodeType('export_statement', true)!;
      const starId = JavaScript.idForNodeType('*', false)!;

      expect(exportStatementId).toBeLessThan(JavaScript.nodeTypeCount);
      expect(starId).toBeLessThan(JavaScript.nodeTypeCount);
      expect(JavaScript.nodeTypeIsNamed(exportStatementId)).toBe(true);
      expect(JavaScript.nodeTypeForId(exportStatementId)).toBe('export_statement');
      expect(JavaScript.nodeTypeIsNamed(starId)).toBe(false);
      expect(JavaScript.nodeTypeForId(starId)).toBe('*');
    });

    it('handles invalid inputs', () => {
      expect(JavaScript.nodeTypeForId(-3)).toBeNull();
      expect(JavaScript.nodeTypeForId(10000)).toBeNull();
      expect(JavaScript.idForNodeType('export_statement', false)).toBeNull();
    });
  });

  describe('Supertypes', () => {
    it('gets the supertypes and subtypes of a parser', () => {
      const supertypes = Rust.supertypes;
      const names = supertypes.map((id) => Rust.nodeTypeForId(id));
      expect(names).toEqual([
        '_expression',
        '_literal',
        '_literal_pattern',
        '_pattern',
        '_type'
      ]);

      for (const id of supertypes) {
        const name = Rust.nodeTypeForId(id);
        const subtypes = Rust.subtypes(id);
        let subtypeNames = subtypes.map((id) => Rust.nodeTypeForId(id));
        subtypeNames = [...new Set(subtypeNames)].sort(); // Remove duplicates & sort

        switch (name) {
          case '_literal':
            expect(subtypeNames).toEqual([
              'boolean_literal',
              'char_literal',
              'float_literal',
              'integer_literal',
              'raw_string_literal',
              'string_literal',
            ]);
            break;
          case '_pattern':
            expect(subtypeNames).toEqual([
              '_',
              '_literal_pattern',
              'captured_pattern',
              'const_block',
              'generic_pattern',
              'identifier',
              'macro_invocation',
              'mut_pattern',
              'or_pattern',
              'range_pattern',
              'ref_pattern',
              'reference_pattern',
              'remaining_field_pattern',
              'scoped_identifier',
              'slice_pattern',
              'struct_pattern',
              'tuple_pattern',
              'tuple_struct_pattern',
            ]);
            break;
          case '_type':
            expect(subtypeNames).toEqual([
              'abstract_type',
              'array_type',
              'bounded_type',
              'dynamic_type',
              'function_type',
              'generic_type',
              'macro_invocation',
              'metavariable',
              'never_type',
              'pointer_type',
              'primitive_type',
              'reference_type',
              'removed_trait_bound',
              'scoped_type_identifier',
              'tuple_type',
              'type_identifier',
              'unit_type',
            ]);
            break;
        }
      }
    });
  });
});

describe('Lookahead iterator', () => {
  let lookahead: LookaheadIterator;
  let state: number;

  beforeAll(async () => {
    ({ JavaScript } = await helper);
    const parser = new Parser();
    parser.setLanguage(JavaScript);
    const tree = parser.parse('function fn() {}')!;
    parser.delete();
    const cursor = tree.walk();
    expect(cursor.gotoFirstChild()).toBe(true);
    expect(cursor.gotoFirstChild()).toBe(true);
    state = cursor.currentNode.nextParseState;
    lookahead = JavaScript.lookaheadIterator(state)!;
    expect(lookahead).toBeDefined();
  });

  afterAll(() => { lookahead.delete() });

  const expected = ['(', 'identifier', '*', 'formal_parameters', 'html_comment', 'comment'];

  it('should iterate over valid symbols in the state', () => {
    const symbols = Array.from(lookahead);
    expect(symbols).toEqual(expect.arrayContaining(expected));
    expect(symbols).toHaveLength(expected.length);
  });

  it('should reset to the initial state', () => {
    expect(lookahead.resetState(state)).toBe(true);
    const symbols = Array.from(lookahead);
    expect(symbols).toEqual(expect.arrayContaining(expected));
    expect(symbols).toHaveLength(expected.length);
  });

  it('should reset', () => {
    expect(lookahead.reset(JavaScript, state)).toBe(true);
    const symbols = Array.from(lookahead);
    expect(symbols).toEqual(expect.arrayContaining(expected));
    expect(symbols).toHaveLength(expected.length);
  });
});
